From 4d4f3d1988eb9130f9a7f3b33c8573196f2c5bc8 Mon Sep 17 00:00:00 2001
From: Marek Vasut <marex@nabladev.com>
Date: Fri, 18 Apr 2025 21:18:00 +0200
Subject: [PATCH 63/82] net: phy: realtek: Add property to enable SSC

Add the 'realtek,ssc-enable' property to enable Spread Spectrum
Clocking (SSC) on Realtek PHYs that support it.

Upstream-Status: Pending
Signed-off-by: Marek Vasut <marex@nabladev.com>
---
 drivers/net/phy/realtek/realtek_main.c | 36 ++++++++++++++++++++++++++
 1 file changed, 36 insertions(+)

diff --git a/drivers/net/phy/realtek/realtek_main.c b/drivers/net/phy/realtek/realtek_main.c
index 5dcc060669039..9c17fc4f6cbde 100644
--- a/drivers/net/phy/realtek/realtek_main.c
+++ b/drivers/net/phy/realtek/realtek_main.c
@@ -73,6 +73,7 @@
 
 #define RTL8211F_PHYCR2				0x19
 #define RTL8211F_CLKOUT_EN			BIT(0)
+#define RTL8211F_SYSCLK_SSC_EN			BIT(3)
 #define RTL8211F_PHYCR2_PHY_EEE_ENABLE		BIT(5)
 
 #define RTL8211F_INSR_PAGE			0xa43
@@ -98,6 +99,12 @@
 #define RTL8211F_RXCR				0x15
 #define RTL8211F_RX_DELAY			BIT(3)
 
+/* RTL8211F SSC settings */
+#define RTL8211F_SSC_PAGE			0xd09
+#define RTL8211F_SSC_CLKOUT			0x10
+#define RTL8211F_SSC_RXC			0x13
+#define RTL8211F_SSC_SYSCLK			0x17
+
 /* RTL8211F WOL settings */
 #define RTL8211F_WOL_PAGE		0xd8a
 #define RTL8211F_WOL_SETTINGS_EVENTS		16
@@ -169,6 +176,7 @@ struct rtl821x_priv {
 	u16 phycr1;
 	u16 phycr2;
 	bool has_phycr2;
+	bool ssc;
 	struct clk *clk;
 	/* rtl8211f */
 	u16 iner;
@@ -247,6 +255,10 @@ static int rtl821x_probe(struct phy_device *phydev)
 		priv->phycr2 = ret & RTL8211F_CLKOUT_EN;
 		if (of_property_read_bool(dev->of_node, "realtek,clkout-disable"))
 			priv->phycr2 &= ~RTL8211F_CLKOUT_EN;
+
+		priv->ssc = of_property_read_bool(dev->of_node, "realtek,ssc-enable");
+		if (priv->ssc)
+			priv->phycr2 |= RTL8211F_SYSCLK_SSC_EN;
 	}
 
 	phydev->priv = priv;
@@ -640,6 +652,30 @@ static int rtl8211f_config_init(struct phy_device *phydev)
 	priv->phycr2 &= ~RTL8211F_PHYCR2_PHY_EEE_ENABLE;
 	phycr2_mask = RTL8211F_PHYCR2_PHY_EEE_ENABLE | RTL8211F_CLKOUT_EN;
 
+	/* Configure spread spectrum */
+	if (priv->ssc) {
+		ret = phy_modify_paged(phydev, 0x0c44, RTL8211F_SSC_RXC, 0xff00, 0x5f00);
+		if (ret < 0) {
+			dev_err(dev, "RXC SCC configuration failed: %pe\n", ERR_PTR(ret));
+			return ret;
+		}
+
+		ret = phy_modify_paged(phydev, 0x0c44, RTL8211F_SSC_SYSCLK, 0xff00, 0x5f00);
+		if (ret < 0) {
+			dev_err(dev, "SYSCLK SCC configuration failed: %pe\n", ERR_PTR(ret));
+			return ret;
+		}
+
+		ret = phy_modify_paged(phydev, RTL8211F_SSC_PAGE, RTL8211F_SSC_CLKOUT,
+				       0xff00, 0xcf00);
+		if (ret < 0) {
+			dev_err(dev, "CLKOUT SCC configuration failed: %pe\n", ERR_PTR(ret));
+			return ret;
+		}
+
+		phycr2_mask |= RTL8211F_SYSCLK_SSC_EN;
+	}
+
 	ret = phy_modify_paged(phydev, RTL8211F_PHYCR_PAGE,
 			       RTL8211F_PHYCR2, phycr2_mask,
 			       priv->phycr2);
-- 
2.51.0

